crates/astria-sequencer/src/sequence/action.rs

```diff
@@ -50,4 +73,58 @@ impl Action {
 }
 
 #[async_trait::async_trait]
-impl ActionHandler for Action {}
+impl ActionHandler for Action {
+    async fn check_stateful<S: StateReadExt + 'static>(
+        &self,
+        state: &S,
+        from: &Address,
+    ) -> Result<()> {
+        let curr_balance = state
+            .get_account_balance(from)
+            .await
+            .context("failed getting `from` account balance")?;
+        let fee = SEQUENCE_ACTION_FEE_PER_BYTE * self.data.len() as u128;
+
+        ensure!(curr_balance >= fee, "insufficient funds");
+
+        Ok(())
+    }
+
+    fn check_stateless(&self) -> Result<()> {
+        ensure!(
+            !self.chain_id.is_empty(),
+            "cannot have empty chain ID for sequence action",
+        );
+        ensure!(
+            self.chain_id.len() <= MAX_CHAIN_ID_LENGTH,
+            "chain ID cannot be longer than {} bytes",
+            MAX_CHAIN_ID_LENGTH,
+        );
+
+        // TODO: do we want to place a maximum on the size of the data?
+        // https://github.com/astriaorg/astria/issues/222
+        ensure!(
+            !self.data.is_empty(),
+            "cannot have empty data for sequence action"
+        );
+        Ok(())
+    }
+
+    #[instrument(
+        skip_all,
+        fields(
+            from = from.to_string(),
+        )
+    )]
+    async fn execute<S: StateWriteExt>(&self, state: &mut S, from: &Address) -> Result<()> {
+        let fee = SEQUENCE_ACTION_FEE_PER_BYTE * self.data.len() as u128;
```

Comment:
Similar to my comment elsewhere: a utility method + unit tests for it would be preferable over calculating this inline.